flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -0,0 +1,423 @@
+	 * discard the state handle call {@link #removeAndDiscardState(String)}.
+	 *
+	 * @param pathInZooKeeper Path of state handle to remove (expected to start with a '/')
+	 * @param callback        The callback after the operation finishes
+	 * @throws Exception If the ZooKeeper operation fails
+	 */
+	public void remove(String pathInZooKeeper, BackgroundCallback callback) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+		checkNotNull(callback, "Background callback");
+
+		client.delete().deletingChildrenIfNeeded().inBackground(callback).forPath(pathInZooKeeper);
+	}
+
+	/**
+	 * Discards a state handle and removes it from ZooKeeper.
+	 *
+	 * <p>If you only want to remove the state handle in ZooKeeper call {@link #remove(String)}.
+	 *
+	 * @param pathInZooKeeper Path of state handle to discard (expected to start with a '/')
+	 * @throws Exception If the ZooKeeper or state handle operation fails
+	 */
+	public void removeAndDiscardState(String pathInZooKeeper) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+
+		StateHandle<T> stateHandle = get(pathInZooKeeper);
+
+		// Delete the state handle from ZooKeeper first
+		client.delete().deletingChildrenIfNeeded().forPath(pathInZooKeeper);
+
+		// Discard the state handle only after it has been successfully deleted from ZooKeeper.
+		// Otherwise we might enter an illegal state after failures (with a state handle in
+		// ZooKeeper, which has already been discarded).
+		stateHandle.discardState();
+	}
+
+	/**
+	 * Discards all available state handles and removes them from ZooKeeper.
+	 *
+	 * @throws Exception If a ZooKeeper or state handle operation fails
+	 */
+	public void removeAndDiscardAllState() throws Exception {
+		final List<ZooKeeperStateHandle<T>> allStateHandles = getAll();
+
+		ZKPaths.deleteChildren(
+				client.getZookeeperClient().getZooKeeper(),
+				ZKPaths.fixForNamespace(client.getNamespace(), "/"),
+				false);
+
+		// Discard the state handles only after they have been successfully deleted from ZooKeeper.
+		for (ZooKeeperStateHandle<T> stateHandle : allStateHandles) {
+			stateHandle.discardState();
+		}
+	}
+
+	/**
+	 * A {@link StateHandle} with its path in ZooKeeper.
+	 *
+	 * @param <T> Type of state
+	 */
+	public static class ZooKeeperStateHandle<T> implements StateHandle<T> {
+
+		private static final long serialVersionUID = -7289601198657398910L;
+
+		private final StateHandle<T> stateHandle;
+
+		private final String pathInZooKeeper;
+
+		private ZooKeeperStateHandle(StateHandle<T> stateHandle, String pathInZooKeeper) {
+			this.stateHandle = checkNotNull(stateHandle, "State handle");
+			this.pathInZooKeeper = checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+		}
+
+		@Override
+		public T getState(ClassLoader userClassLoader) throws Exception {
+			return stateHandle.getState(userClassLoader);
+		}
+
+		@Override
+		public void discardState() throws Exception {
```

Comment:
But this should be easily doable. Simply call `ZooKeeperStateHandleStore.remove(pathInZooKeeper, callback)` from the `discardState` method in `ZooKeeperStateHandle`. Then it would also be in sync with the `StateHandle` interface that if you call `discard` that the state gets discarded. Only with the difference, that in case of an failure in ZooKeeper, the wrapped `StateHandle` won't be discarded.